### PR TITLE
2.4x avg speedup on PMG struct to CHGNET graph conversion

### PR DESCRIPTION
**Pull request by AegisIK**
Converted `graph` data structure creation (not be confused with the `Crystal_Graph` object) into c (using Cython as an interfacing layer). Still supports legacy graph creation if c-supported graph creation library import fails.

Overall average `graph` creation speedup: 3x
Overall average `Crystal_Graph` creation speedup: 2.4x
Overall average MD speedup (on GPU): ~2x (may be a noisy benchmark however)

Benchmark on Ac2AgPb supercell:

image

### CODE REVIEW

File: chgnet/graph/fast_converter_libraries/uthash.h [hunk 1]
```diff
@@ -0,0 +1,1140 @@
+/*
```
Comment (janosh) [inline chgnet/graph/fast_converter_libraries/uthash.h:1]: Asking without any background knowledge here but do we actually need to check this into version control? I know C is not as batteries-included as Python but seems strange that you even have to bring your own hash table.

Comment (AegisIK) [reply chgnet/graph/fast_converter_libraries/uthash.h:1]: C doesn't support any data structures (unless strings are a data structure) in their standard library -- in fact, their standard library is extremely sparse compared to languages such as Java or Python. This is a useful hashmap I found, implemented purely in C macros, on the interwebs.

Comment (lbluque) [reply chgnet/graph/fast_converter_libraries/uthash.h:1]: Do any of the C++ standard data structures work for this case?

Comment (AegisIK) [reply chgnet/graph/fast_converter_libraries/uthash.h:1]: There might be -- I'm not familiar with C++ however. The uthash.h library is simple to use.

File: pyproject.toml [hunk 2]
```diff
@@ -15,6 +15,7 @@ dependencies = [
     "numpy>=1.21.6",
     "pymatgen>=2022.4.19",
     "torch>=1.11.0",
+    "cython>=0.29.26"
```
Comment (BowenD-UCB) [inline pyproject.toml:18]: @AegisIK @lbluque Should cython be capitalized here as "Cython" ?

Comment (janosh) [reply pyproject.toml:18]: `pip` is case insensitive. Also, hyphens and underscores are equivalent in package names.

**Review by janosh — APPROVED** [review]
Looking great! I think this is ready to go.

**Review by BowenD-UCB — APPROVED** [review]

**Review by BowenD-UCB — APPROVED** [review]

**Pull request merged**